### ISSUE TITLE
[Bug]: HI3 Registry Export Failure

### ISSUE THREAD

**Issue by sino1641** [CollapseLauncher/Collapse#341]
### Affected Version

171 11 PRE

### Bug Behavior

I have noticed FPS unlocks are not working on the latest PRE.
Launcher settings:
<img width="884" alt="Snipaste_2023-12-17_21-54-02" src="https://github.com/CollapseLauncher/Collapse/assets/13870295/8a93982e-d697-4dfe-bca3-105b20dad127">
In-game FPS:
<img width="1280" alt="Snipaste_2023-12-17_21-56-26" src="https://github.com/CollapseLauncher/Collapse/assets/13870295/590c53e0-a563-46cf-8a1c-4c8afd8bb565">

### Expected Behavior

Unlock FPS to expecting values, eg 120.
It works on previous versions, but I don't know it fails in which update specifically.

### Steps to reproduce

1. Launch the game.
2. Go to any combat.

### Additional Information

Launcher logs:
[log-2023-12-17-pre1.72.11-id1.log](https://github.com/CollapseLauncher/Collapse/files/13696343/log-2023-12-17-pre1.72.11-id1.log)

**Comment by bagusnl** [comment]
Eyyo~~
Thank you for submitting bug report to Collapse.

The behavior is strange, since we have not change anything about HI3 Game Settings for quite a while. Can you tell us which region is this?

Tested on mine SEA 7.0 works fine on both 100 and 120.
![image](https://github.com/CollapseLauncher/Collapse/assets/28079733/5db031c6-a855-4ec1-a3ff-a4b6dbef7c9a)

**Comment by sino1641** [comment]
Hi bagusnl~ Thanks for testing!
I am playing 7.1 CN.
And the good news is I can set it up to 144 now after I changed FPS settings in the game.
It might be come conflicts occurred before.

But I tried to export registry with following error:
```Shell
[Schm]  Page changed to GameSettingsPage with Tag: gamesettings
[Info]  Exporting registry Software\miHoYo\崩坏3...
[Info]  Writing UnitySelectMonitor_h17969598...
[Info]  Writing Screenmanager Resolution Width_h182942802...
[Info]  Writing Screenmanager Resolution Height_h2627697771...
[Info]  Writing Screenmanager Is Fullscreen mode_h3981298716...
[Info]  Writing UnityGraphicsQuality_h1669003810...
[Info]  Writing unity.player_session_count_h922449978...
[Info]  Writing unity.player_sessionid_h1351336811...
[Info]  Writing MIHOYOSDK_DEVICE_FP_h1619094146...
[Info]  Writing GENERAL_DATA_V2_LastLoginServer_h100100692...
[Info]  Writing GENERAL_DATA_V2_UserLocalDataVersion_h4195919598...
[Info]  Writing GENERAL_DATA_V2_LastSelectedRegionName_h1506518084...
[Info]  Writing GENERAL_DATA_V2_IsConvertedFromPersonalGraphicsSetting_h2689903257...
[Info]  Writing MIHOYOSDK_CURRENT_LANGUAGE_h2559149783...
[Info]  Writing SDK_LANG_VERSION_zh-cn_0_h1790562469...
[Info]  Writing SDK_LANG_FILE_zh-cn_0_h2597224953...
[Info]  Writing MIHOYOSDK_PROTOCOL__zh-cn_h1189219383...
[Info]  Writing MIHOYOSDK_USER_AGREEMENT_SHOW_FLAG_zh-cn_h466595801...
[Info]  Writing MIHOYOSDK_PROTOCOL_SHOW_FLAG__zh-cn_h656742520...
[Info]  Writing MIHOYOSDK_ANNOUNCE_URL_h25198164...
[Info]  Writing MIHOYOSDK_CONFIG_MODEL_NAME_h2176277873...
[Info]  Writing MIHOYOSDK_SEED_ID_h2996644406...
[Info]  Writing MIHOYOSDK_SEED_TIME_h3465600238...
[Info]  Writing GENERAL_DATA_V2_LastBindingSourceType_h2972011318...
[Info]  Writing GENERAL_DATA_V2_LastDeviceStyle_h4242067005...
[Info]  Writing MiHoYoSDKPreLoadEnable_h2423167308...
[Info]  Writing GENERAL_DATA_V2_ResourceDownloadType_h2238376574...
[Info]  Writing GENERAL_DATA_V2_ResourceDownloadVersion_h1528433916...
[Info]  Writing MIHOYOSDK_ADL_PROD_CN_h3123967166...
[Info]  Writing GENERAL_DATA_V2_Account_h3411181689...
[Info]  Writing MIHOYOSDK_PROTOCOL_1_74861998_zh-cn_h2126890717...
[Info]  Writing USD_V2_266251153_LastWeatherInfo_h904464243...
[Info]  Writing USD_V2_266251153_PlayerSpeedTimeCurrentKeyName_h3147443511...
[Info]  Writing USD_V2_266251153_PlayerSpeedTimeDataBeginTimestamp_h3288090423...
[Info]  Writing USD_V2_266251153_LastPlayerMainlineMissionClickData_h2108658145...
[Info]  Writing USD_V2_266251153_LastPlayerSpeedTimeDataDict_h3269678830...
[Info]  Writing USD_V2_266251153_ReportedSystemInfo_h372038795...
[Info]  Writing USD_V2_266251153_TeamDict_h2126591880...
[Info]  Writing USD_V2_266251153_LastTraceMainlineStepID_h1482296894...
[Info]  Writing GENERAL_DATA_V2_LastLoginUserId_h47158221...
[Info]  Writing GENERAL_DATA_V2_LastLoginUserActive_h843017996...
[Info]  Writing GENERAL_DATA_V2_LastLoginAccountUID_h3749284904...
[Info]  Writing GENERAL_DATA_V2_AccountLastLoginServerDict_h1988878063...
[Info]  Writing USD_V2_821625356_AvatarOverviewFavourate_h3468357885...
[Info]  Writing USD_V2_821625356_PlayerSpeedTimeCurrentKeyName_h3470870206...
[Info]  Writing USD_V2_821625356_PlayerSpeedTimeDataBeginTimestamp_h846147006...
[Info]  Writing USD_V2_821625356_LastPlayerSpeedTimeDataDict_h3760734823...
[Info]  Writing USD_V2_821625356_ReportedSystemInfo_h329407842...
[Info]  Writing USD_V2_821625356_LastWeatherInfo_h3072429882...
[Info]  Writing USD_V2_821625356_TeamDict_h3360043169...
[Info]  Writing USD_V2_821625356_LastTraceMainlineStepID_h2379605303...
[Info]  Writing USD_V2_821625356_OpenworldNewMapList_h961138066...
[Info]  Writing USD_V2_821625356_LocalSinDemonBeginTime_h555110400...
[Info]  Writing USD_V2_821625356_GrandKeyWeaponRecord_h1108550208...
[Info]  Writing USD_V2_821625356_VersionPVPlayRecordMap_h2714210191...
[Info]  Writing USD_V2_821625356_LocalMedalCache_h3019746470...
[Info]  Writing USD_V2_821625356_LocalEmojiCache_h3086964771...
[Info]  Writing USD_V2_821625356_RpgTaleLocalDataV2_h1240996262...
[Info]  Writing USD_V2_821625356_DLCTowerTeam_h63280011...
[Info]  Writing USD_V2_821625356_DLCTowerTeamEverFilled_h126490145...
[Info]  Writing USD_V2_821625356_SeenChapterActivitySection_h497344294...
[Info]  Writing USD_V2_821625356_EnteredMainLineChapterIds_h1499251693...
[Info]  Writing USD_V2_821625356_LocalEndlessBeginTime_h1913652895...
[Info]  Writing USD_V2_821625356_WaveRushScheduleID_h40740951...
[Info]  Writing USD_V2_821625356_HaveEnterOpenwroldQuest_h1657287268...
[Info]  Writing USD_V2_821625356_LastOpenworldRefreshTime_h2904979730...
[Info]  Writing USD_V2_821625356_LatestShortStoryExModeDict_h1145858984...
[Info]  Writing USD_V2_821625356_LatestShortStoryExDifficultyDict_h3015362392...
[Info]  Writing USD_V2_821625356_MonthlyCardLastExpireHintTimeStamp_h3262988955...
[Info]  Writing USD_V2_821625356_PlayerSpeedTimeDataDict_h514829293...
[Info]  Writing USD_V2_821625356_PlayerSpeedTimeCurrentKeyEnterTimestamp_h319541759...
[Info]  Writing USD_V2_821625356_LastShowBulletinTime_h1549124539...
[Info]  Writing USD_V2_821625356_LocalVersion_h330236401...
[Info]  Writing USD_V2_821625356_ActivityRemindTimeStamp_h1493829436...
[Info]  Writing USD_V2_821625356_TeamDictTrialStage_h3750650791...
[Info]  Writing USD_V2_821625356_LastPlayerLoginTimestamp_h4198479122...
[Info]  Writing USD_V2_821625356_StageSwitchTrialAvatarUsingDic_h3283959275...
[Info]  Writing USD_V2_821625356_WaveRushTeamDict_h3116855416...
[Info]  Writing USD_V2_821625356_WaveRushElfDict_h3521710666...
[Info]  Writing USD_V2_821625356_WaveRushDefaultTeam_h2808569353...
[Info]  Writing USD_V2_821625356_WaveRushDefaultElf_h3338887707...
[Info]  Writing USD_V2_821625356_AvatarAppearanceCount_h2454267018...
[Info]  Writing MIHOYO_SDK_FATIGUE_REMINDER_TOTAL_TIME_bh3_cn_74861998_h2220596776...
[Info]  Writing MIHOYO_SDK_FATIGUE_REMINDER_RESET_DATE_bh3_cn_74861998_h2596386494...
[Info]  Writing MIHOYO_SDK_FATIGUE_REMINDER_ACTIVE_TIMESTAMP_bh3_cn_74861998_h1179272253...
[Info]  Writing USD_V2_821625356_OldMailCache_h2644311716...
[Info]  Writing USD_V2_821625356_CheckedNewFixBoxList_h1731123613...
[Info]  Writing USD_V2_821625356_LastWorldMapFeatureType_h2232249918...
[Info]  Writing USD_V2_821625356_OWAvatarTeamDict_h2417300232...
[Info]  Writing USD_V2_821625356_LastLevelType_h2330336066...
[Info]  Writing USD_V2_821625356_OpenWorldTeamDict_h308734007...
[Info]  Writing USD_V2_821625356_LastWorldMapFeatureSelectedContentID_h1452203819...
[Info]  Writing USD_V2_821625356_MPRaidRefreshHint_h3631193411...
[Info]  Writing USD_V2_821625356_MPQuickJoinLevel_h597317290...
[Info]  Writing USD_V2_821625356_MPAvatarID_h2610362023...
[Info]  Writing USD_V2_821625356_MPStrangerInfoList_h3404213117...
[Info]  Writing USD_V2_821625356_MPLevelLastUsedAvatarIDsDict_h3769834773...
[Info]  Writing USD_V2_821625356_MpLevelLastUsedElfIDDictionary_h2480017978...
[Info]  Writing USD_V2_821625356_NewWantedElfDict_h2281445698...
[Info]  Writing USD_V2_821625356_NewWantedTeamDict_h1673698416...
[Info]  Writing USD_V2_821625356_LastShowOpenWorldServantPopUpTimeForLowLife_h1547350086...
[Info]  Writing USD_V2_821625356_LastShowALUpHintLevel_h4055699722...
[Info]  Writing USD_V2_821625356_LastShowAdventureExplore_h2808429944...
[Info]  Writing USD_V2_821625356_IsFirstTimeEnterOpenWorldQuestSys_h1633041778...
[Info]  Writing USD_V2_821625356_DoneBehaviourList_h204008413...
[Info]  Writing USD_V2_821625356_ElfTeamDict_h1116766382...
[Info]  Writing USD_V2_821625356_OpenWorldJudgeScoreLocal_h3268032908...
[Erro]  Error has occurred while exporting registry!
System.InvalidCastException: Unable to cast object of type 'System.Int64' to type 'System.Int32'.
   at CollapseLauncher.Pages.GameSettingsPage.RegistryExportClick(Object sender, RoutedEventArgs e)
```
Should I raise another issue or just post it here?

**Comment by Cryotechnic** [comment]
Hi, no need for making another issue as it is related to the problem above. We'll make it and ping you from there if need be. Currently under investigation, we'll keep you posted of any updates. Just want to confirm though, did this issue go away like the FPS issue or is it still persisting?

**Comment by bagusnl** [comment]
The issue is in `WriteDWord(EndianBinaryWriter writer, object obj)` in `ImportExportBase.cs`. Value for certain keys is for some reason a Int64 rather than the DWORD default Int32 (HOYOOOOOOOOOO), this causes a cast error and break the whole exporter.

While the writer issue is easily fixed by changing the cast to `(int)(long)` when `InvalidCastException` is raised, the importer has no idea that this has happened and caused the value of the key to be zeroed out after importing the clreg.

Need assistance from @neon-nyan

**Comment by sino1641** [comment]
@Cryotechnic Yes the export failure still exists. 
I am not pros but from @bagusnl It seems a great effort to fix this data type conversion error in both import and export functions.

**Comment by Cryotechnic** [comment]
I believe cf7802e addresses the issue. We'll have a new release soon so please look forward to that. Keeping this open until it is deployed for tracking.

**Comment by bagusnl** [comment]
The fix has been published in Preview 1.72.12, closing this issue. 

Thank you for reporting the issue!
